File: src/pixel_art_mcp/imaging/preview.py
"""Selected, nearest-neighbor asset previews delivered directly through MCP."""

import io
import json
from pathlib import Path
from typing import Any

from PIL import Image, ImageOps

from pixel_art_mcp.imaging.context import context_image
from pixel_art_mcp.imaging.inspection import _selection
from pixel_art_mcp.models import DomainError


def asset_preview(
    root: Path,
    clip_id: str | None,
    angle: int | None,
    frame: int | None,
    scale: int,
    context: bool,
) -> tuple[bytes, dict[str, Any]]:
    if not 1 <= scale <= 8:
        raise DomainError("Preview scale must be an integer from 1 to 8")
    metadata = json.loads((root / "spritesheet.json").read_text())
    kind = metadata["asset"]["kind"]
    clip_id = clip_id or next(iter(metadata["asset"]["clips"]))
    requested_angle = metadata["layouts"][0]["angle"] if angle is None else angle
    source_angle = requested_angle
    mirrored = kind != "furniture" and requested_angle == 270
    if mirrored:
        source_angle = 90
    if kind == "pet" and clip_id == "idle" and requested_angle in (90, 270):
        source_angle = 0 if requested_angle == 90 else 180
        mirrored = False
    metadata, entry, path, _ = _selection(root, clip_id, source_angle, frame)
    with Image.open(path) as opened:
        sprite = opened.convert("RGBA")
    if mirrored:
        sprite = ImageOps.mirror(sprite)
    native_size = list(sprite.size)
    layout = next(row for row in metadata["layouts"] if row["angle"] == source_angle)
    if context:
        sprite = context_image(sprite, metadata["asset"], layout)
    size = (sprite.width * scale, sprite.height * scale)
    if size[0] * size[1] > 4_194_304:
        raise DomainError("Preview exceeds 4194304 pixels; choose a smaller scale")
    sprite = sprite.resize(size, Image.Resampling.NEAREST)
    stream = io.BytesIO()
    sprite.save(stream, "PNG")
    return stream.getvalue(), {
        "clip_id": clip_id,
        "angle": requested_angle,
        "source_angle": source_angle,
        "frame": entry["frame"],
        "mirrored": mirrored,
        "native_size": native_size,
        "display_size": list(size),
        "scale": scale,
        "context": context,
        "context_is_approximate": context,
        "visual_review_required": True,
    }

File: packages/imaging/src/preview.ts
/**
 * Selected, nearest-neighbor asset previews delivered directly through MCP. Verbatim port of
 * `src/pixel_art_mcp/imaging/preview.py` (63 lines).
 */

import fs from "node:fs";
import path from "node:path";

import { DomainError, type FurnitureCategory } from "@pixel-art-mcp/schema";

import { contextImage, type ContextLayout } from "./context.js";
import { encodePngBuffer, readPng, resizeNearest, type RGBAImage } from "./image.js";
import { selectSpriteFrame } from "./inspection.js";
import { defined } from "./internal.js";

/** The subset of an asset's `spritesheet.json` `assetPreview` reads. */
export interface AssetPreviewMetadata {
  asset: {
    kind: "furniture" | "character" | "pet";
    category: FurnitureCategory;
    placement: "floor" | "surface" | "wall";
    clips: Record<string, unknown>;
  };
  layouts: readonly ContextLayout[];
  [key: string]: unknown;
}

export interface AssetPreviewResult {
  clip_id: string;
  angle: number;
  source_angle: number;
  frame: number;
  mirrored: boolean;
  native_size: [number, number];
  display_size: [number, number];
  scale: number;
  context: boolean;
  context_is_approximate: boolean;
  visual_review_required: true;
}

/** `Image.open(...).convert("RGBA")` with a horizontal mirror, matching PIL's `ImageOps.mirror`
 * (flips columns left-right; alpha carried through unchanged). */
function mirrorHorizontal(image: RGBAImage): RGBAImage {
  const out = { width: image.width, height: image.height, data: new Uint8Array(image.data.length) };
  for (let y = 0; y < image.height; y++) {
    for (let x = 0; x < image.width; x++) {
      const srcOffset = (y * image.width + x) * 4;
      const dstOffset = (y * image.width + (image.width - 1 - x)) * 4;
      out.data[dstOffset] = defined(image.data[srcOffset], "mirror pixel");
      out.data[dstOffset + 1] = defined(image.data[srcOffset + 1], "mirror pixel");
      out.data[dstOffset + 2] = defined(image.data[srcOffset + 2], "mirror pixel");
      out.data[dstOffset + 3] = defined(image.data[srcOffset + 3], "mirror pixel");
    }
  }
  return out;
}

export function assetPreview(
  root: string,
  clipId: string | null,
  angle: number | null,
  frame: number | null,
  scale: number,
  context: boolean,
): [Buffer, AssetPreviewResult] {
  if (!Number.isInteger(scale) || scale < 1 || scale > 8) {
    throw new DomainError("Preview scale must be an integer from 1 to 8");
  }
  const raw = JSON.parse(fs.readFileSync(path.join(root, "spritesheet.json"), "utf-8")) as AssetPreviewMetadata;
  const kind = raw.asset.kind;
  const resolvedClipId = clipId ?? defined(Object.keys(raw.asset.clips)[0], "first clip id");
  const firstLayout = defined(raw.layouts[0], "first layout");
  const requestedAngle = angle ?? firstLayout.angle;
  let sourceAngle = requestedAngle;
  let mirrored = kind !== "furniture" && requestedAngle === 270;
  if (mirrored) sourceAngle = 90;
  if (kind === "pet" && resolvedClipId === "idle" && (requestedAngle === 90 || requestedAngle === 270)) {
    sourceAngle = requestedAngle === 90 ? 0 : 180;
    mirrored = false;
  }

  const { entry, filePath } = selectSpriteFrame(root, resolvedClipId, sourceAngle, frame);
  let sprite = readPng(filePath);
  if (mirrored) sprite = mirrorHorizontal(sprite);
  const nativeSize: [number, number] = [sprite.width, sprite.height];
  const layout = defined(
    raw.layouts.find((row) => row.angle === sourceAngle),
    `layout for angle ${String(sourceAngle)}`,
  );
  let stage = sprite;
  if (context) {
    stage = contextImage(sprite, raw.asset, layout);
  }
  const size: [number, number] = [stage.width * scale, stage.height * scale];
  if (size[0] * size[1] > 4_194_304) {
    throw new DomainError("Preview exceeds 4194304 pixels; choose a smaller scale");
  }
  const resized = resizeNearest(stage, size[0], size[1]);
  const buffer = encodePngBuffer(resized);
  return [
    buffer,
    {
      clip_id: resolvedClipId,
      angle: requestedAngle,
      source_angle: sourceAngle,
      frame: entry.frame,
      mirrored,
      native_size: nativeSize,
      display_size: size,
      scale,
      context,
      context_is_approximate: context,
      visual_review_required: true,
    },
  ];
}
